### PR TITLE
chore: http/2 benchmark

### PR DESCRIPTION
**Pull request by mkaufmaner**
Start the server:
```bash
PORT=3000 node --experimental-wasm-simd ./benchmarks/server-http2.js
```

Run the benchmark:
```bash
PORT=3000 ERROR_THRESHOLD=100 CONNECTIONS=10 PARALLEL=10 node --experimental-wasm-simd ./benchmarks/benchmark-http2.js 
```

**_Note: May have to run `benchmark-http2.js` multiple times, and in quick succession, in order to get a result._**

Result:
```
│ Tests             │ Samples │          Result │ Tolerance │ Difference with slowest │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ http2 - request   │       1 │  449.75 req/sec │  ± 0.00 % │                       - │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - fetch    │       1 │  509.34 req/sec │  ± 0.00 % │               + 13.25 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - request  │       1 │ 1086.15 req/sec │  ± 0.00 % │              + 141.50 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - stream   │       1 │ 1168.15 req/sec │  ± 0.00 % │              + 159.73 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - pipeline │       1 │ 1200.61 req/sec │  ± 0.00 % │              + 166.95 % │
```

### CODE REVIEW

**Comment by metcoder95** [comment]
The results indeed looks promising, though if the streams are better managed, the things can go even better as we'll be reusing them to the same origin.
Let me fix some stuff to make it work with `dispatch` and focus on the performance improvements later on. Great work btw 🚀

**Comment by metcoder95** [comment]
Hey @mkaufmaner, can you rebase with the branch and enable the benchmarks?
Feel free to open a new PR if the conflicts are complex to resolve 🙂

**Comment by mkaufmaner** [comment]
> Hey @mkaufmaner, can you rebase with the branch and enable the benchmarks? Feel free to open a new PR if the conflicts are complex to resolve 🙂

Merged down from your http2 branch, but I am not sure what you mean by "enable the benchmarks". Could you elaborate?

Secondly, I updated the benchmarks with `allowH2: true` and reran them. Still facing the intermittent issue of them failing though.

```
│ Tests             │ Samples │          Result │ Tolerance │ Difference with slowest │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - fetch    │       1 │  464.24 req/sec │  ± 0.00 % │                       - │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ http2 - request   │       1 │  464.87 req/sec │  ± 0.00 % │                + 0.14 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - pipeline │       1 │ 1143.19 req/sec │  ± 0.00 % │              + 146.25 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - request  │       1 │ 1255.94 req/sec │  ± 0.00 % │              + 170.54 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - stream   │       1 │ 1364.41 req/sec │  ± 0.00 % │              + 193.90 % │
```

**Comment by mkaufmaner** [comment]
@metcoder95 
> Hey @mkaufmaner, can you rebase with the branch and enable the benchmarks? Feel free to open a new PR if the conflicts are complex to resolve 🙂

> Hey @mkaufmaner, can you rebase with the branch and enable the benchmarks? Feel free to open a new PR if the conflicts are complex to resolve 🙂

Merged down from your http2 branch, but I am not sure what you mean by "enable the benchmarks". Could you elaborate?

Secondly, I updated the benchmarks with `allowH2: true`, enabled the `dispatch` benchmark, and reran the http2 benchmark. I am still encountering the same issue before where sometimes the benchmark just doesn't print anything so you have to run it in succession to get a result.

Nonetheless, 
```
│ Tests             │ Samples │          Result │ Tolerance │ Difference with slowest │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ http2 - request   │       1 │  459.30 req/sec │  ± 0.00 % │                       - │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - fetch    │       1 │  518.96 req/sec │  ± 0.00 % │               + 12.99 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - request  │       1 │  951.29 req/sec │  ± 0.00 % │              + 107.12 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - pipeline │       1 │ 1116.78 req/sec │  ± 0.00 % │              + 143.15 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - stream   │       1 │ 1169.42 req/sec │  ± 0.00 % │              + 154.61 % │
|───────────────────|─────────|─────────────────|───────────|─────────────────────────|
│ undici - dispatch │       1 │ 1230.43 req/sec │  ± 0.00 % │              + 167.89 % │
```

File: benchmarks/benchmark-http2.js [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+            write (chunk, encoding, callback) {
+              callback()
+            }
+          })
+        )
+        .on('finish', resolve)
+    })
+  },
+  'undici - request' () {
+    // console.log('undici - request')
```
Comment (metcoder95) [inline benchmarks/benchmark-http2.js:228]: left over?

File: benchmarks/server-http2.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+      setTimeout(() => {
+        // res.send(buf)
+        // res.end()
+        res.end(buf)
+      }, timeout)
+    }
+  )
+
+  // let streamErrorCount = 0;
```
Comment (metcoder95) [inline benchmarks/server-http2.js:55]: left over?

**Review by metcoder95 — COMMENTED** [review]
>Could you elaborate?

Sure, I was meaning those that were commented out

>Secondly, I updated the benchmarks with allowH2: true, enabled the dispatch benchmark, and reran the http2 benchmark. I am still encountering the same issue before where sometimes the benchmark just doesn't print anything so you have to run it in succession to get a result.

Those are good numbers, thanks! 🙇 
I'm not sure what the error might be, I'm guessing that is due to how the request is done, as they differ from the way `undici` usually works, although not really important if we can get the numbers


Would you like to go further and do the cleanup to merge?

**Comment by metcoder95** [comment]
Linter seems failing :(

**Comment by mkaufmaner** [comment]
> Linter seems failing :(

Fixed the lint errors.

**Pull request merged**